security/pkg/k8s/chiron/controller.go

```diff
@@ -0,0 +1,818 @@
+	}
+
+	// The service ports of the webhooks. Each item corresponds to an item at the same index in WebhookServiceNames.
+	WebhookServicePorts = []int{
+		// This is a service port created as a demo of the prototype
+		443,
+		// TODO (lei-tang): remove demo service port and enable the following webhook service ports
+		//	443,
+		//	443,
+	}
+)
+
+// WebhookController manages the service accounts' secrets that contains Istio keys and certificates.
+type WebhookController struct {
+	certTTL        time.Duration
+	k8sClient      *kubernetes.Clientset
+	core           corev1.CoreV1Interface
+	minGracePeriod time.Duration
+	// Length of the grace period for the certificate rotation.
+	gracePeriodRatio float32
+
+	// Whether the certificates are for dual-use clients (SAN+CN).
+	dualUse bool
+
+	// If true, generate a PKCS#8 private key.
+	pkcs8Key bool
+
+	// DNS-enabled serviceAccount.namespace to service pair
+	dnsNames map[string]*crl.DNSNameEntry
+
+	// Controller and store for service account objects.
+	saController cache.Controller
+	saStore      cache.Store
+
+	// Controller and store for secret objects.
+	scrtController cache.Controller
+	scrtStore      cache.Store
+
+	monitoring monitoringMetrics
+
+	certClient certclient.CertificatesV1beta1Interface
+
+	// The namespace of the webhook certificates
+	namespace string
+
+	// MutatingWebhookConfiguration
+	mutatingWebhookConfigName string
+	// The name of the webhook in the MutatingWebhookConfiguration
+	mutatingWebhookName string
+
+	// Watcher for the CA certificate
+	CaCertWatcher *fsnotify.Watcher
+
+	// Current CA certificate
+	curCACert []byte
+
+	mutex sync.RWMutex
+}
+
+// NewWebhookController returns a pointer to a newly constructed WebhookController instance.
+func NewWebhookController(certTTL time.Duration,
+	gracePeriodRatio float32, minGracePeriod time.Duration, dualUse bool, k8sClient *kubernetes.Clientset,
+	core corev1.CoreV1Interface, certClient certclient.CertificatesV1beta1Interface, pkcs8Key bool,
+	dnsNames map[string]*crl.DNSNameEntry, nameSpace, mutatingWebhookConfigName, mutatingWebhookName string) (*WebhookController, error) {
+
+	if gracePeriodRatio < 0 || gracePeriodRatio > 1 {
+		return nil, fmt.Errorf("grace period ratio %f should be within [0, 1]", gracePeriodRatio)
+	}
+	if gracePeriodRatio < recommendedMinGracePeriodRatio || gracePeriodRatio > recommendedMaxGracePeriodRatio {
+		log.Warnf("grace period ratio %f is out of the recommended window [%.2f, %.2f]",
+			gracePeriodRatio, recommendedMinGracePeriodRatio, recommendedMaxGracePeriodRatio)
+	}
+
+	c := &WebhookController{
+		certTTL:                   certTTL,
+		gracePeriodRatio:          gracePeriodRatio,
+		minGracePeriod:            minGracePeriod,
+		dualUse:                   dualUse,
+		k8sClient:                 k8sClient,
+		core:                      core,
+		pkcs8Key:                  pkcs8Key,
+		dnsNames:                  dnsNames,
+		monitoring:                newMonitoringMetrics(),
```

Comment:
Do we want to have the monitoring enabled in the first version?

Reply:
Monitoring is removed.